examples/notebooks/blogposts/unittests_in_beam.ipynb

```diff
@@ -0,0 +1,259 @@
+        "# specific language governing permissions and limitations\n",
+        "# under the License"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "# Install the Apache Beam library\n",
+        "\n",
+        "!pip install apache_beam[gcp] --quiet"
+      ],
+      "metadata": {
+        "id": "5W2nuV7uzlPg"
+      },
+      "execution_count": 37,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#The following packages are used to run the example pipelines\n",
+        "\n",
+        "import apache_beam as beam\n",
+        "from apache_beam.io import ReadFromText, WriteToText\n",
+        "from apache_beam.options.pipeline_options import PipelineOptions\n",
+        "\n",
+        "class CustomClass(beam.DoFn):\n",
+        "  def custom_function(x):\n",
+        "          ...\n",
+        "          # returned_record = requests.get(\"http://my-api-call.com\")\n",
+        "          ...\n",
+        "          # if len(returned_record)!=10:\n",
+        "          # raise ValueError(\"Length of record does not match expected length\")\n",
+        "          return x\n",
+        "\n",
+        "  with beam.Pipeline() as p:\n",
+        "    result = (\n",
+        "            p\n",
+        "            | ReadFromText(\"/content/sample_data/anscombe.json\")\n",
+        "            | beam.ParDo(lambda x: CustomClass.custom_function(x))\n",
+        "            | WriteToText(\"/content/\")\n",
+        "    )"
+      ],
+      "metadata": {
+        "id": "Ktk9EVIFzGfP"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "**Example Pipeline 1**\n"
+      ],
+      "metadata": {
+        "id": "IVjBkewt1sLA"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "# This function is going to return the square the integer at the first index of our record.\n",
+        "def compute_square(element):\n",
+        "  return int(element[1])**2\n",
+        "\n",
+        "with beam.Pipeline() as p1:\n",
+        "    result = (\n",
+        "        p1\n",
+        "        | ReadFromText(\"/content/sample_data/california_housing_test.csv\",skip_header_lines=1)\n",
+        "        | beam.Map(compute_square)\n",
+        "        | WriteToText(\"/content/\")\n",
+        "    )"
+      ],
+      "metadata": {
+        "id": "oHbSvOUI1pOe"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "**Example Pipeline 2**"
+      ],
+      "metadata": {
+        "id": "Mh3nZZ1_12sX"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "with beam.Pipeline() as p2:\n",
+        "    result = (\n",
+        "        p2\n",
+        "        | ReadFromText(\"/content/sample_data/anscombe.json\")\n",
+        "        | beam.Map(str.strip)\n",
```

Comment:
I wonder if a better way to do this would be to split `beam.Map(str.strip)` out into a separate function which can be called from the test. As it is, the test isn't actually invoking any of the code we've written.

A more interesting example might be:

```
def manipulate_strings(incoming_pcoll):
   return incoming_pcoll | beam.Map(str.strip) | beam.Map(str.upper)
```

The functions themselves don't need tested, but the beam transforms do. That would let you test the actual code you've written below with:

```
  with TestPipeline() as p:
    inputs = p | beam.Create(strings)
    output = manipulate_strings(inputs)
    assert_that(output, equal_to(expected))
```

Reply:
Basically, I don't like that you could totally change the user code (which you're supposed to be testing) without a test failing.

Reply:
Seeing `output = manipulate_strings(inputs)` makes me wonder how we could make composite PTransforms even easier/more natural.

Reply:
> Basically, I don't like that you could totally change the user code (which you're supposed to be testing) without a test failing.

I see your point and think it's valid (I can make the change). Out of curiosity though, what if a users entire transform was an inbuilt function (like `str.strip`)? Would the guidance be that they wouldn't need to test the Beam transform?